In [2]:
import pandas as pd

In [3]:
year = [2010,2011,2012,2013,2014,2015]

salaryHCM = [8.5,9.0,10.5,12.0,11.0,10.0]

salaryLA = [5.5,6.0,6.0,5.5,6.0,5.0]

df = pd.DataFrame(list(zip(year, salaryHCM,salaryLA)),
                  columns=['Year', 'HCM','LA'])
df

Year   HCM   LA
0  2010   8.5  5.5
1  2011   9.0  6.0
2  2012  10.5  6.0
3  2013  12.0  5.5
4  2014  11.0  6.0
5  2015  10.0  5.0

MaxAbsolute scale và đoạn [-1;1]

In [4]:
max_la = df['LA'].abs().max()
df['macLA'] = df['LA'] / max_la

df

Year   HCM   LA     macLA
0  2010   8.5  5.5  0.916667
1  2011   9.0  6.0  1.000000
2  2012  10.5  6.0  1.000000
3  2013  12.0  5.5  0.916667
4  2014  11.0  6.0  1.000000
5  2015  10.0  5.0  0.833333

In [5]:
max_HCM = df['HCM'].abs().max()
df['macHCM'] = df['HCM'] / max_HCM

df

Year   HCM   LA     macLA    macHCM
0  2010   8.5  5.5  0.916667  0.708333
1  2011   9.0  6.0  1.000000  0.750000
2  2012  10.5  6.0  1.000000  0.875000
3  2013  12.0  5.5  0.916667  1.000000
4  2014  11.0  6.0  1.000000  0.916667
5  2015  10.0  5.0  0.833333  0.833333

In [6]:
from sklearn.preprocessing import MaxAbsScaler
scaler = MaxAbsScaler()
scaled = scaler.fit_transform(df[['HCM','LA']])
scaled_df = pd.DataFrame(scaled, columns=df[['HCM','LA']].columns)

print(scaled_df)

        HCM        LA
0  0.708333  0.916667
1  0.750000  1.000000
2  0.875000  1.000000
3  1.000000  0.916667
4  0.916667  1.000000
5  0.833333  0.833333


HÃY CHO BIẾT VỚI LƯƠNG HCM VÀ LA LẦN LƯỢT LÀ 4.5 VÀ 5.75 

In [7]:
vSc = scaler.transform([[4.5,5.75]])
vSc

d:\Python\Py_Data\Lib\site-packages\sklearn\base.py:493: UserWarning: X does not have valid feature names, but MaxAbsScaler was fitted with feature names
  warnings.warn(


array([[0.375     , 0.95833333]])

HÃY CHO BIẾT VỚI LƯƠNG HCM VÀ LA TRƯỚC KHI SCALE LÀ BAO NHIEU NẾU BIẾT TRC GIA TRỊ TRƯỚC KHI SCALE LÀ 0.5 VÀ 0.8

In [8]:
vR = scaler.inverse_transform([[0.5,0.8]])
vR

array([[6. , 4.8]])

# MIN_MAX_SCALER

Xnew = (Xstd - min) / (max - min) thuộc [0;1]

In [9]:
min_la = df['LA'].min()
max_la = df['LA'].max()
df['mmcLA'] = (df['LA'] - min_la) / (max_la - min_la)

df

Year   HCM   LA     macLA    macHCM  mmcLA
0  2010   8.5  5.5  0.916667  0.708333    0.5
1  2011   9.0  6.0  1.000000  0.750000    1.0
2  2012  10.5  6.0  1.000000  0.875000    1.0
3  2013  12.0  5.5  0.916667  1.000000    0.5
4  2014  11.0  6.0  1.000000  0.916667    1.0
5  2015  10.0  5.0  0.833333  0.833333    0.0

In [10]:
min_hcm = df['HCM'].min()
max_hcm = df['HCM'].max()
df['mmchcm'] = (df['HCM'] - min_hcm) / (max_hcm - min_hcm)

df

Year   HCM   LA     macLA    macHCM  mmcLA    mmchcm
0  2010   8.5  5.5  0.916667  0.708333    0.5  0.000000
1  2011   9.0  6.0  1.000000  0.750000    1.0  0.142857
2  2012  10.5  6.0  1.000000  0.875000    1.0  0.571429
3  2013  12.0  5.5  0.916667  1.000000    0.5  1.000000
4  2014  11.0  6.0  1.000000  0.916667    1.0  0.714286
5  2015  10.0  5.0  0.833333  0.833333    0.0  0.428571

In [11]:
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()
scaled = scaler.fit_transform(df[['HCM','LA']])
scaled_df = pd.DataFrame(scaled, columns=df[['HCM','LA']].columns)

print(scaled_df)

        HCM   LA
0  0.000000  0.5
1  0.142857  1.0
2  0.571429  1.0
3  1.000000  0.5
4  0.714286  1.0
5  0.428571  0.0


Giả sử lương hcm 10tr LA 4.5tr sau khi qua minmaxscale là bao nhieu

In [12]:
(4.5-5) / (6-4.5)

-0.3333333333333333

# Z-SCORE

In [13]:
df[['HCM','LA']].describe()

HCM        LA
count   6.000000  6.000000
mean   10.166667  5.666667
std     1.290994  0.408248
min     8.500000  5.000000
25%     9.250000  5.500000
50%    10.250000  5.750000
75%    10.875000  6.000000
max    12.000000  6.000000

In [14]:
#Zhcm
Zhcm = (8.5 - 10.17) / 1.29
Zhcm

-1.2945736434108526

In [15]:
#Zla
Zla= (5.5 -5.67)/0.4
Zla

-0.4249999999999998

std HCM = 1.290994 cách quá xa o vớ Zhcm = -1.29 còn std LA = 0.4 cách không xa Zla = -0.42 nên ở LONGAN thoải mái hon

# Nếu 1 người có thu nhập ở hcm là 15tr và LA là 12tr thì ở đâu thoải mái hơn

In [16]:
mean_la = df['LA'].mean()
std_la = df['LA'].std()
df['zLA'] = (df['LA'] - mean_la) / std_la

df

Year   HCM   LA     macLA    macHCM  mmcLA    mmchcm       zLA
0  2010   8.5  5.5  0.916667  0.708333    0.5  0.000000 -0.408248
1  2011   9.0  6.0  1.000000  0.750000    1.0  0.142857  0.816497
2  2012  10.5  6.0  1.000000  0.875000    1.0  0.571429  0.816497
3  2013  12.0  5.5  0.916667  1.000000    0.5  1.000000 -0.408248
4  2014  11.0  6.0  1.000000  0.916667    1.0  0.714286  0.816497
5  2015  10.0  5.0  0.833333  0.833333    0.0  0.428571 -1.632993

In [17]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
scaled = scaler.fit_transform(df[['HCM','LA']])

scaled_df = pd.DataFrame(scaled, columns=df[['HCM','LA']].columns)

print(scaled_df)

        HCM        LA
0 -1.414214 -0.447214
1 -0.989949  0.894427
2  0.282843  0.894427
3  1.555635 -0.447214
4  0.707107  0.894427
5 -0.141421 -1.788854


In [18]:
sV = scaler.transform([[15,12]])
sV


d:\Python\Py_Data\Lib\site-packages\sklearn\base.py:493: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  warnings.warn(


array([[ 4.10121933, 16.99411663]])

# function

In [19]:
def funcX(x):
    return x*x + 1

df['TransformX'] = funcX(df[["LA"]])

df

Year   HCM   LA     macLA    macHCM  mmcLA    mmchcm       zLA  TransformX
0  2010   8.5  5.5  0.916667  0.708333    0.5  0.000000 -0.408248       31.25
1  2011   9.0  6.0  1.000000  0.750000    1.0  0.142857  0.816497       37.00
2  2012  10.5  6.0  1.000000  0.875000    1.0  0.571429  0.816497       37.00
3  2013  12.0  5.5  0.916667  1.000000    0.5  1.000000 -0.408248       31.25
4  2014  11.0  6.0  1.000000  0.916667    1.0  0.714286  0.816497       37.00
5  2015  10.0  5.0  0.833333  0.833333    0.0  0.428571 -1.632993       26.00